File: gtci/03-sliding-window/python3/8-string-anagrams.py
import math


class Solution:
    def findStringAnagrams(self, str1, pattern):
        result_indices = []

        pattern_chars = self.construct_freq_map_from_string(pattern)

        chars = {}
        windowStart = 0
        for windowEnd in range(len(str1)):
            windowSize = windowEnd - windowStart + 1

            if str1[windowEnd] not in chars:
                chars[str1[windowEnd]] = 0
            chars[str1[windowEnd]] += 1

            while windowSize > len(pattern):
                chars[str1[windowStart]] -= 1
                if chars[str1[windowStart]] == 0:
                    del chars[str1[windowStart]]

                windowStart += 1
                windowSize = windowEnd - windowStart + 1

            if windowSize == len(pattern) and chars == pattern_chars:
                result_indices.append(windowStart)

        return result_indices

    def construct_freq_map_from_string(self, s):
        chars = {}
        for char in s:
            if char not in chars:
                chars[char] = 0
            chars[char] += 1

        return chars


if __name__ == "__main__":
    sol = Solution()

    assert sol.findStringAnagrams("ppqp", "pq") == [1, 2]
    assert sol.findStringAnagrams("abbcabc", "abc") == [2, 3, 4]

    print("All test cases passed.")

File: gtci/03-sliding-window/cpp/8-string-anagrams.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<int> findStringAnagrams(const string &str, const string &pattern) {
        vector<int> resultIndices;

        std::unordered_map<char, int> pattern_chars =
            Solution::constructFreqMapFromString(pattern);

        std::unordered_map<char, int> chars;
        int windowStart = 0;
        for (int windowEnd = 0; windowEnd < str.size(); windowEnd++) {
            int windowSize = windowEnd - windowStart + 1;

            chars[str[windowEnd]] += 1;

            while (windowSize > pattern.size()) {
                chars[str[windowStart]] -= 1;
                if (chars[str[windowStart]] == 0) chars.erase(str[windowStart]);

                windowStart++;
                windowSize = windowEnd - windowStart + 1;
            }

            if (windowSize == pattern.size() && chars == pattern_chars)
                resultIndices.push_back(windowStart);
        }

        return resultIndices;
    }

   private:
    static std::unordered_map<char, int> constructFreqMapFromString(
        const std::string &str) {
        std::unordered_map<char, int> chars;

        for (char ch : str) chars[ch] += 1;

        return chars;
    }
};

int main() {
    Solution sol;

    assert(sol.findStringAnagrams("ppqp", "pq") == vector<int>({1, 2}));
    assert(sol.findStringAnagrams("abbcabc", "abc") == vector<int>({2, 3, 4}));

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
